In [1]:
# =============================================================================
# STEP 1: LOAD PROCESSED DATASET
# =============================================================================

import pandas as pd
import numpy as np


# Load cleaned dataset
df = pd.read_csv(
    "../../data/processed/breast_cancer_cleaned.csv"
)


print("=" * 70)
print("PROCESSED DATASET")
print("=" * 70)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")


display(df.head())

PROCESSED DATASET
Rows    : 569
Columns : 31


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


In [2]:
# =============================================================================
# STEP 2: FEATURE AND TARGET SEPARATION
# =============================================================================


target_column = "target"


X = df.drop(
    target_column,
    axis=1
)


y = df[target_column]



print("=" * 70)
print("FEATURE MATRIX (X)")
print("=" * 70)

print(f"Rows    : {X.shape[0]}")
print(f"Columns : {X.shape[1]}")


print("\n" + "=" * 70)
print("TARGET VARIABLE (y)")
print("=" * 70)

print(f"Rows : {y.shape[0]}")

display(y.head())

FEATURE MATRIX (X)
Rows    : 569
Columns : 30

TARGET VARIABLE (y)
Rows : 569


0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64

In [3]:
# =============================================================================
# STEP 3: VERIFY TARGET DISTRIBUTION
# =============================================================================


print("=" * 70)
print("TARGET DISTRIBUTION")
print("=" * 70)


display(
    y.value_counts()
)


print("\nClass Percentage")


display(
    (y.value_counts(normalize=True)*100)
    .round(2)
)

TARGET DISTRIBUTION


target
1    357
0    212
Name: count, dtype: int64


Class Percentage


target
1    62.74
0    37.26
Name: proportion, dtype: float64

In [4]:
# =============================================================================
# STEP 4: FEATURE TYPE CHECK
# =============================================================================


print("=" * 70)
print("FEATURE DATA TYPES")
print("=" * 70)


display(
    X.dtypes.to_frame(
        name="Data Type"
    )
)

FEATURE DATA TYPES


,Data Type
mean radius,float64
mean texture,float64
mean perimeter,float64
mean area,float64
mean smoothness,float64
mean compactness,float64
mean concavity,float64
mean concave points,float64
mean symmetry,float64
mean fractal dimension,float64


In [5]:
# =============================================================================
# STEP 5: MISSING VALUE CHECK
# =============================================================================
# This step is included to document the data-quality check explicitly.
# No imputation is performed here. If the processed dataset is complete,
# the feature matrix is carried forward unchanged.

print("=" * 70)
print("MISSING VALUES")
print("=" * 70)

missing = pd.DataFrame({
    "Missing Count": X.isnull().sum(),
    "Missing Percentage (%)": (
        X.isnull().mean() * 100
    ).round(2)
})

display(missing[missing["Missing Count"] > 0])

total_missing = int(X.isnull().sum().sum())

print("\nTotal missing values:", total_missing)

if total_missing == 0:
    print("No missing values detected. No imputation is required.")
else:
    raise ValueError(
        "Missing values were detected. "
        "Stop before feature selection and handle missingness explicitly."
    )


MISSING VALUES


,Missing Count,Missing Percentage (%)



Total missing values: 0
No missing values detected. No imputation is required.


In [6]:
# =============================================================================
# STEP 6: TRAIN-TEST SPLIT
# =============================================================================


from sklearn.model_selection import train_test_split


print("=" * 70)
print("TRAIN-TEST SPLIT")
print("=" * 70)


X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)


print("Training Data")
print("-" * 70)

print(f"X_train shape : {X_train.shape}")
print(f"y_train shape : {y_train.shape}")


print("\nTesting Data")
print("-" * 70)

print(f"X_test shape  : {X_test.shape}")
print(f"y_test shape  : {y_test.shape}")

TRAIN-TEST SPLIT
Training Data
----------------------------------------------------------------------
X_train shape : (455, 30)
y_train shape : (455,)

Testing Data
----------------------------------------------------------------------
X_test shape  : (114, 30)
y_test shape  : (114,)


In [7]:
# =============================================================================
# STEP 7: CLASS DISTRIBUTION AFTER SPLIT
# =============================================================================


print("=" * 70)
print("TRAINING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_train.value_counts(normalize=True)*100)
    .round(2)
)



print("\n" + "=" * 70)
print("TESTING TARGET DISTRIBUTION")
print("=" * 70)


display(
    (y_test.value_counts(normalize=True)*100)
    .round(2)
)

TRAINING TARGET DISTRIBUTION


target
1    62.64
0    37.36
Name: proportion, dtype: float64


TESTING TARGET DISTRIBUTION


target
1    63.16
0    36.84
Name: proportion, dtype: float64

In [8]:
# =============================================================================
# STEP 8: FEATURE SCALING
# =============================================================================


from sklearn.preprocessing import StandardScaler


print("=" * 70)
print("FEATURE SCALING USING STANDARD SCALER")
print("=" * 70)


# Initialize scaler
scaler = StandardScaler()


# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)


# Transform test data using same scaler
X_test_scaled = scaler.transform(X_test)



print("Scaling completed successfully")


print("\n" + "=" * 70)
print("SHAPE AFTER SCALING")
print("=" * 70)


print(f"X_train_scaled shape : {X_train_scaled.shape}")
print(f"X_test_scaled shape  : {X_test_scaled.shape}")

FEATURE SCALING USING STANDARD SCALER
Scaling completed successfully

SHAPE AFTER SCALING
X_train_scaled shape : (455, 30)
X_test_scaled shape  : (114, 30)


In [9]:
# =============================================================================
# CONVERT SCALED DATA BACK TO DATAFRAME
# =============================================================================


X_train_scaled = pd.DataFrame(
    X_train_scaled,
    columns=X_train.columns,
    index=X_train.index
)


X_test_scaled = pd.DataFrame(
    X_test_scaled,
    columns=X_test.columns,
    index=X_test.index
)



print("=" * 70)
print("SCALED TRAIN DATA PREVIEW")
print("=" * 70)


display(X_train_scaled.head())

SCALED TRAIN DATA PREVIEW


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst radius,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension
546,-1.072001,-0.658425,-1.088080,-0.939274,-0.135940,-1.008718,-0.968359,-1.102032,0.281062,-0.113231,...,-1.034094,-0.623497,-1.070773,-0.876534,-0.169982,-1.038836,-1.078995,-1.350527,-0.352658,-0.541380
432,1.748743,0.066502,1.751157,1.745559,1.274468,0.842288,1.519852,1.994664,-0.293045,-0.320180,...,1.228342,-0.092833,1.187467,1.104386,1.517001,0.249655,1.178594,1.549916,0.191078,-0.173739
174,-0.974734,-0.931124,-0.997709,-0.867589,-0.613515,-1.138154,-1.092292,-1.243358,0.434395,-0.429247,...,-0.973231,-1.036772,-1.008044,-0.834168,-1.097823,-1.167260,-1.282241,-1.707442,-0.307734,-1.213033
221,-0.145103,-1.215186,-0.123013,-0.253192,0.664482,0.286762,-0.129729,-0.098605,0.555635,0.029395,...,-0.251266,-1.369643,-0.166633,-0.330292,0.234006,0.096874,-0.087521,-0.344838,0.242198,-0.118266
289,-0.771617,-0.081211,-0.803700,-0.732927,-0.672282,-1.006099,-0.798502,-0.684484,0.737495,-0.457213,...,-0.801135,0.079230,-0.824381,-0.741830,-0.911367,-0.984612,-0.933190,-0.777604,0.555118,-0.761639


In [10]:
# =============================================================================
# SCALING VERIFICATION
# =============================================================================


print("=" * 70)
print("MEAN AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.mean().round(2)
)



print("\n" + "=" * 70)
print("STANDARD DEVIATION AFTER SCALING")
print("=" * 70)

display(
    X_train_scaled.std().round(2)
)

MEAN AFTER SCALING


mean radius               -0.0
mean texture               0.0
mean perimeter            -0.0
mean area                 -0.0
mean smoothness            0.0
mean compactness          -0.0
mean concavity            -0.0
mean concave points        0.0
mean symmetry             -0.0
mean fractal dimension     0.0
radius error              -0.0
texture error             -0.0
perimeter error            0.0
area error                 0.0
smoothness error          -0.0
compactness error         -0.0
concavity error            0.0
concave points error      -0.0
symmetry error             0.0
fractal dimension error    0.0
worst radius               0.0
worst texture             -0.0
worst perimeter           -0.0
worst area                 0.0
worst smoothness           0.0
worst compactness         -0.0
worst concavity            0.0
worst concave points      -0.0
worst symmetry            -0.0
worst fractal dimension    0.0
dtype: float64


STANDARD DEVIATION AFTER SCALING


mean radius                1.0
mean texture               1.0
mean perimeter             1.0
mean area                  1.0
mean smoothness            1.0
mean compactness           1.0
mean concavity             1.0
mean concave points        1.0
mean symmetry              1.0
mean fractal dimension     1.0
radius error               1.0
texture error              1.0
perimeter error            1.0
area error                 1.0
smoothness error           1.0
compactness error          1.0
concavity error            1.0
concave points error       1.0
symmetry error             1.0
fractal dimension error    1.0
worst radius               1.0
worst texture              1.0
worst perimeter            1.0
worst area                 1.0
worst smoothness           1.0
worst compactness          1.0
worst concavity            1.0
worst concave points       1.0
worst symmetry             1.0
worst fractal dimension    1.0
dtype: float64

# ANOVA

## 1. Baseline

In [11]:
# =============================================================================
# ANOVA BASELINE FEATURE SELECTION (TOP-K = 5,10,15,20)
# =============================================================================

from sklearn.feature_selection import SelectKBest, f_classif
import pandas as pd


# Fit ANOVA once with all features
anova_selector = SelectKBest(
    score_func=f_classif,

    k="all"
)

anova_selector.fit(
    X_train_scaled,
    y_train
)


# Create score dataframe
anova_scores = pd.DataFrame({

    "Feature": X_train_scaled.columns,

    "ANOVA Score": anova_selector.scores_

})


# Sort by ANOVA importance
anova_scores = anova_scores.sort_values(
    by="ANOVA Score",
    ascending=False
).reset_index(drop=True)



# Top-k values
k_values = [5, 10, 15, 20]

anova_results = {}


for k in k_values:

    top_features = (
        anova_scores
        .head(k)["Feature"]
        .tolist()
    )


    # Boolean mask for selecting columns
    mask = X_train_scaled.columns.isin(
        top_features
    )


    X_train_selected = X_train_scaled.loc[
        :, mask
    ]

    X_test_selected = X_test_scaled.loc[
        :, mask
    ]


    anova_results[k] = {

        "features": top_features,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "scores": anova_scores.head(k)

    }


    print("="*70)
    print(f"TOP {k} ANOVA FEATURES")
    print("="*70)

    print(top_features)

    print("\nScores:")
    print(anova_scores.head(k))

TOP 5 ANOVA FEATURES
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter']

Scores:
                Feature  ANOVA Score
0  worst concave points   795.223341
1       worst perimeter   719.598823
2   mean concave points   703.364031
3          worst radius   687.430367
4        mean perimeter   565.485755
TOP 10 ANOVA FEATURES
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter', 'worst area', 'mean radius', 'mean area', 'mean concavity', 'worst concavity']

Scores:
                Feature  ANOVA Score
0  worst concave points   795.223341
1       worst perimeter   719.598823
2   mean concave points   703.364031
3          worst radius   687.430367
4        mean perimeter   565.485755
5            worst area   541.713694
6           mean radius   523.613907
7             mean area   469.516453
8        mean concavity   416.374874
9       worst concavity   365.747389
TOP 15 ANOVA FEATURES
['worst

In [12]:
# =============================================================================
# BASELINE MODELS
# =============================================================================

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier


models = {
    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=42
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=42
    ),

    "XGBoost": XGBClassifier(
        random_state=42,

        eval_metric="logloss",
        scale_pos_weight=(y_train == 0).sum() / (y_train == 1).sum()
    )
}

In [13]:
# =============================================================================
# MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

results = []

for k in k_values:

    Xtr = anova_results[k]["X_train"]
    Xts = anova_results[k]["X_test"]

    features = ", ".join(anova_results[k]["features"])

    for model_name, model in models.items():

        model.fit(Xtr, y_train)

        y_pred = model.predict(Xts)

        y_prob = model.predict_proba(Xts)[:,1]

        results.append({

            "Method": "ANOVA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(y_test, y_pred),

            "Precision": precision_score(y_test, y_pred),

            "Recall": recall_score(y_test, y_pred),

            "F1 Score": f1_score(y_test, y_pred),

            "ROC-AUC": roc_auc_score(y_test, y_prob)

        })

In [14]:
results_df = pd.DataFrame(results)

results_df

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,ANOVA,5,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.929825,0.984848,0.902778,0.942029,0.991071
1,ANOVA,5,Random Forest,"worst concave points, worst perimeter, mean co...",0.929825,0.957143,0.930556,0.943662,0.988261
2,ANOVA,5,XGBoost,"worst concave points, worst perimeter, mean co...",0.938596,0.957746,0.944444,0.951049,0.984458
3,ANOVA,10,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.912281,0.969697,0.888889,0.927536,0.991402
4,ANOVA,10,Random Forest,"worst concave points, worst perimeter, mean co...",0.956140,0.958904,0.972222,0.965517,0.990079
5,ANOVA,10,XGBoost,"worst concave points, worst perimeter, mean co...",0.973684,0.972603,0.986111,0.979310,0.992394
6,ANOVA,15,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.956140,0.985507,0.944444,0.964539,0.991402
7,ANOVA,15,Random Forest,"worst concave points, worst perimeter, mean co...",0.956140,0.971831,0.958333,0.965035,0.991733
8,ANOVA,15,XGBoost,"worst concave points, worst perimeter, mean co...",0.964912,0.959459,0.986111,0.972603,0.992394
9,ANOVA,20,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.964912,0.985714,0.958333,0.971831,0.995040


In [16]:
results_df.to_csv(
    "../../../results/breast_cancer/baseline_results/anova_baseline_results.csv",
    index=False
)

## 2. DODA

In [17]:
import sys
import subprocess

subprocess.check_call([
    sys.executable,
    "-m",
    "pip",
    "install",
    "--upgrade",
    "--force-reinstall",
    "--no-cache-dir",
    "git+https://github.com/anandha-3679/DODA.git"
])

0

In [18]:
import sys
import doda
from doda import DODASelector
import inspect

print("Python:", sys.executable)
print("DODA:", doda.__file__)
print("Selector:", inspect.getfile(DODASelector))

Python: c:\Users\johnm\msc_research\.venv\Scripts\python.exe
DODA: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\__init__.py
Selector: c:\Users\johnm\msc_research\.venv\Lib\site-packages\doda\selector.py


In [19]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

from doda import DODASelector

from doda.adapters import SklearnAdapter

from doda.knowledge.providers import JSONProvider

from doda.fusion import HadamardFusion

In [20]:
# =============================================================================
# ANOVA + DODA FEATURE SELECTION
# =============================================================================

import pandas as pd
from sklearn.feature_selection import SelectKBest, f_classif

# -------------------------------------------------------------------------
# Top-K values
# -------------------------------------------------------------------------

k_values = [5, 10, 15, 20]


# -------------------------------------------------------------------------
# Clinical knowledge provider
# -------------------------------------------------------------------------

provider = JSONProvider(
    "../../../config/clinical_weights/breast_cancer.json"
)


# -------------------------------------------------------------------------
# Fusion method
# -------------------------------------------------------------------------

fusion = HadamardFusion()


# -------------------------------------------------------------------------
# Store DODA results
# -------------------------------------------------------------------------

doda_results = {}


# =============================================================================
# RUN DODA FOR EACH TOP-K
# =============================================================================

for k in k_values:

    print("\n" + "=" * 70)
    print(f"ANOVA + DODA \u2014 TOP {k} FEATURES")
    print("=" * 70)


    # ---------------------------------------------------------------------
    # ANOVA operator
    # ---------------------------------------------------------------------

    anova = SklearnAdapter(
        SelectKBest(
            score_func=f_classif,
            k=k
        )
    )


    # ---------------------------------------------------------------------
    # DODA selector
    # ---------------------------------------------------------------------

    selector = DODASelector(
        operators=[anova],
        provider=provider,
        fusion=fusion,
        top_k=k
    )


    # ---------------------------------------------------------------------
    # Fit DODA on TRAINING data
    # ---------------------------------------------------------------------

    X_train_selected = selector.fit_transform(
        X_train_scaled,
        y_train
    )


    # ---------------------------------------------------------------------
    # Transform TEST data
    # ---------------------------------------------------------------------

    X_test_selected = selector.transform(
        X_test_scaled
    )


    # ---------------------------------------------------------------------
    # Selected features
    # ---------------------------------------------------------------------

    selected_features = selector.get_selected_features()


    print("\nSelected Features:")
    print(selected_features)


    # ---------------------------------------------------------------------
    # Store everything needed for later evaluation
    # ---------------------------------------------------------------------

    doda_results[k] = {

        "selector": selector,

        "X_train": X_train_selected,

        "X_test": X_test_selected,

        "features": selected_features,

        "raw_math_scores": selector.raw_math_scores_,

        "math_scores": selector.math_scores_,

        "clinical_weights": selector.clinical_weights_,

        "final_scores": selector.final_scores_
    }


print("\n" + "=" * 70)
print("DODA FEATURE SELECTION COMPLETED")
print("=" * 70)


ANOVA + DODA — TOP 5 FEATURES
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x000001A13DACD2B0>
DODA Pipeline Started
Running: SelectKBest
DODA feature order:
['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension']
Number of features: 30
Number of scores: 30
Operator Scores:
{'SelectKBest_1': {'mean radius': np.float64(523.6139074486017), 'mean texture': np.float64(91.20625420859241), 'mean perimeter': np.float64(565.4857549199432), 'mean area': np.float64(469.5164529562025), 'mean smoothness': np.float64(82.34753778975492), 'mean compactness': np.float64(269.46604199334104), 'mean concavity': np.float64(416.37487400408787), 'mean concave points': np.float64(703.3640310829362), 'mean symmetry': np.float64(66.75130382442448), 'mean fractal dimension': np.float64(0.00047821117736673776), 'radius error': np.fl

In [21]:
# CELL 2
# =============================================================================
# INSPECT MATHEMATICAL SCORES
# =============================================================================

k = 10

selector = doda_results[k]["selector"]

# -----------------------------------------------------------------------------
# Raw mathematical scores
# -----------------------------------------------------------------------------

raw_math_scores = pd.DataFrame(
    selector.raw_math_scores_.items(),
    columns=[
        "Feature",
        "Raw Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Normalized mathematical scores
# -----------------------------------------------------------------------------

normalized_math_scores = pd.DataFrame(
    selector.math_scores_.items(),
    columns=[
        "Feature",
        "Normalized Math Score"
    ]
)

# -----------------------------------------------------------------------------
# Combine both
# -----------------------------------------------------------------------------

math_scores = raw_math_scores.merge(
    normalized_math_scores,
    on="Feature"
)

math_scores = math_scores.sort_values(
    by="Normalized Math Score",
    ascending=False
)

print("=" * 70)
print("MATHEMATICAL SCORES")
print("=" * 70)

display(math_scores)

MATHEMATICAL SCORES


,Feature,Raw Math Score,Normalized Math Score
27,worst concave points,795.223341,1.000000e+00
22,worst perimeter,719.598823,9.049015e-01
7,mean concave points,703.364031,8.844861e-01
20,worst radius,687.430367,8.644494e-01
2,mean perimeter,565.485755,7.111031e-01
23,worst area,541.713694,6.812095e-01
0,mean radius,523.613907,6.584489e-01
3,mean area,469.516453,5.904209e-01
6,mean concavity,416.374874,5.235949e-01
26,worst concavity,365.747389,4.599304e-01


In [22]:
# CELL 3
# =============================================================================
# INSPECT CLINICAL WEIGHTS
# =============================================================================

clinical_weights = pd.DataFrame(
    selector.clinical_weights_.items(),
    columns=[
        "Feature",
        "Clinical Weight"
    ]
)

clinical_weights = clinical_weights.sort_values(
    by="Clinical Weight",
    ascending=False
)

print("=" * 70)
print("CLINICAL WEIGHTS")
print("=" * 70)

display(clinical_weights)

CLINICAL WEIGHTS


,Feature,Clinical Weight
27,worst concave points,1.0
26,worst concavity,1.0
22,worst perimeter,1.0
23,worst area,1.0
20,worst radius,1.0
0,mean radius,0.9
3,mean area,0.9
6,mean concavity,0.9
7,mean concave points,0.9
2,mean perimeter,0.9


In [23]:
# CELL 4
# =============================================================================
# INSPECT FINAL DODA SCORES
# =============================================================================

final_scores = pd.DataFrame(
    selector.final_scores_.items(),
    columns=[
        "Feature",
        "Final DODA Score"
    ]
)

final_scores = final_scores.sort_values(
    by="Final DODA Score",
    ascending=False
)

print("=" * 70)
print("FINAL DODA SCORES")
print("=" * 70)

display(final_scores)

FINAL DODA SCORES


,Feature,Final DODA Score
27,worst concave points,1.000000e+00
22,worst perimeter,9.049015e-01
20,worst radius,8.644494e-01
7,mean concave points,7.960375e-01
23,worst area,6.812095e-01
2,mean perimeter,6.399928e-01
0,mean radius,5.926040e-01
3,mean area,5.313788e-01
6,mean concavity,4.712354e-01
26,worst concavity,4.599304e-01


In [24]:
selector.get_selected_features()

['worst concave points',
 'worst perimeter',
 'worst radius',
 'mean concave points',
 'worst area',
 'mean perimeter',
 'mean radius',
 'mean area',
 'mean concavity',
 'worst concavity']

In [25]:
comparison = selector.compare_scores()

comparison


MATHEMATICAL vs CLINICAL vs HADAMARDFUSION

Fusion Method: HadamardFusion

Top Features After Fusion:
                Feature  Math Rank  Clinical Rank  Final Rank  Rank Change
0  worst concave points          1              5           1            0
1       worst perimeter          2              2           2            0
2          worst radius          4              1           3            1
3   mean concave points          3             10           4           -1
4            worst area          6              3           5            1
5        mean perimeter          5              7           6           -1
6           mean radius          7              6           7            0
7             mean area          8              8           8            0
8        mean concavity          9              9           9            0
9       worst concavity         10              4          10            0


,Feature,Math Rank,Normalized Math Score,Clinical Rank,Clinical Weight,Final Rank,Final Score,Rank Change
0,worst concave points,1,1.0000,5,1.0,1,1.0000,0
1,worst perimeter,2,0.9049,2,1.0,2,0.9049,0
2,worst radius,4,0.8644,1,1.0,3,0.8644,1
3,mean concave points,3,0.8845,10,0.9,4,0.7960,-1
4,worst area,6,0.6812,3,1.0,5,0.6812,1
5,mean perimeter,5,0.7111,7,0.9,6,0.6400,-1
6,mean radius,7,0.6584,6,0.9,7,0.5926,0
7,mean area,8,0.5904,8,0.9,8,0.5314,0
8,mean concavity,9,0.5236,9,0.9,9,0.4712,0
9,worst concavity,10,0.4599,4,1.0,10,0.4599,0


In [26]:
# =============================================================================
# DODA MODEL EVALUATION
# =============================================================================

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)


doda_model_results = []


for k in k_values:

    Xtr = doda_results[k]["X_train"]

    Xts = doda_results[k]["X_test"]

    features = ", ".join(
        doda_results[k]["features"]
    )


    for model_name, model in models.items():

        model.fit(
            Xtr,
            y_train
        )


        y_pred = model.predict(
            Xts
        )


        y_prob = model.predict_proba(
            Xts
        )[:, 1]


        doda_model_results.append({

            "Method": "ANOVA + DODA",

            "Top-K": k,

            "Model": model_name,

            "Selected Features": features,

            "Accuracy": accuracy_score(
                y_test,
                y_pred
            ),

            "Precision": precision_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "Recall": recall_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "F1 Score": f1_score(
                y_test,
                y_pred,
                zero_division=0
            ),

            "ROC-AUC": roc_auc_score(
                y_test,
                y_prob
            )
        })


doda_results_df = pd.DataFrame(
    doda_model_results
)


display(doda_results_df)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,ANOVA + DODA,5,Logistic Regression,"worst concave points, worst perimeter, worst r...",0.938596,0.985075,0.916667,0.949640,0.990741
1,ANOVA + DODA,5,Random Forest,"worst concave points, worst perimeter, worst r...",0.938596,0.971014,0.930556,0.950355,0.982804
2,ANOVA + DODA,5,XGBoost,"worst concave points, worst perimeter, worst r...",0.921053,0.943662,0.930556,0.937063,0.982474
3,ANOVA + DODA,10,Logistic Regression,"worst concave points, worst perimeter, worst r...",0.912281,0.969697,0.888889,0.927536,0.991402
4,ANOVA + DODA,10,Random Forest,"worst concave points, worst perimeter, worst r...",0.947368,0.958333,0.958333,0.958333,0.989253
5,ANOVA + DODA,10,XGBoost,"worst concave points, worst perimeter, worst r...",0.964912,0.959459,0.986111,0.972603,0.991071
6,ANOVA + DODA,15,Logistic Regression,"worst concave points, worst perimeter, worst r...",0.956140,0.985507,0.944444,0.964539,0.991402
7,ANOVA + DODA,15,Random Forest,"worst concave points, worst perimeter, worst r...",0.964912,0.972222,0.972222,0.972222,0.989087
8,ANOVA + DODA,15,XGBoost,"worst concave points, worst perimeter, worst r...",0.964912,0.959459,0.986111,0.972603,0.992725
9,ANOVA + DODA,20,Logistic Regression,"worst concave points, worst perimeter, worst r...",0.964912,0.985714,0.958333,0.971831,0.995040


In [27]:
doda_results_df.to_csv(
    "../../../results/breast_cancer/doda_results/anova_doda_results.csv",
    index=False
)

In [28]:
# =============================================================================
# COMBINED COMPARISON
# =============================================================================

comparison_df = pd.concat(
    [
        results_df,
        doda_results_df
    ],
    ignore_index=True
)


display(
    comparison_df
)

,Method,Top-K,Model,Selected Features,Accuracy,Precision,Recall,F1 Score,ROC-AUC
0,ANOVA,5,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.929825,0.984848,0.902778,0.942029,0.991071
1,ANOVA,5,Random Forest,"worst concave points, worst perimeter, mean co...",0.929825,0.957143,0.930556,0.943662,0.988261
2,ANOVA,5,XGBoost,"worst concave points, worst perimeter, mean co...",0.938596,0.957746,0.944444,0.951049,0.984458
3,ANOVA,10,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.912281,0.969697,0.888889,0.927536,0.991402
4,ANOVA,10,Random Forest,"worst concave points, worst perimeter, mean co...",0.956140,0.958904,0.972222,0.965517,0.990079
5,ANOVA,10,XGBoost,"worst concave points, worst perimeter, mean co...",0.973684,0.972603,0.986111,0.979310,0.992394
6,ANOVA,15,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.956140,0.985507,0.944444,0.964539,0.991402
7,ANOVA,15,Random Forest,"worst concave points, worst perimeter, mean co...",0.956140,0.971831,0.958333,0.965035,0.991733
8,ANOVA,15,XGBoost,"worst concave points, worst perimeter, mean co...",0.964912,0.959459,0.986111,0.972603,0.992394
9,ANOVA,20,Logistic Regression,"worst concave points, worst perimeter, mean co...",0.964912,0.985714,0.958333,0.971831,0.995040


In [29]:
for k in k_values:

    selector = doda_results[k]["selector"]

    raw_rank = [
        x[0]
        for x in sorted(
            selector.raw_math_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    doda_rank = [
        x[0]
        for x in sorted(
            selector.final_scores_.items(),
            key=lambda x: x[1],
            reverse=True
        )
    ]

    print("\n" + "=" * 70)
    print(f"TOP {k}")
    print("=" * 70)

    print("Raw Math Ranking:")
    print(raw_rank)

    print("\nDODA Ranking:")
    print(doda_rank)

    print("\nTop-K Same:",
          raw_rank[:k] == doda_rank[:k])


TOP 5
Raw Math Ranking:
['worst concave points', 'worst perimeter', 'mean concave points', 'worst radius', 'mean perimeter', 'worst area', 'mean radius', 'mean area', 'mean concavity', 'worst concavity', 'mean compactness', 'worst compactness', 'area error', 'radius error', 'perimeter error', 'worst texture', 'worst symmetry', 'worst smoothness', 'concave points error', 'mean texture', 'mean smoothness', 'mean symmetry', 'worst fractal dimension', 'compactness error', 'concavity error', 'smoothness error', 'fractal dimension error', 'texture error', 'symmetry error', 'mean fractal dimension']

DODA Ranking:
['worst concave points', 'worst perimeter', 'worst radius', 'mean concave points', 'worst area', 'mean perimeter', 'mean radius', 'mean area', 'mean concavity', 'worst concavity', 'worst compactness', 'mean compactness', 'area error', 'radius error', 'perimeter error', 'worst texture', 'worst symmetry', 'worst smoothness', 'mean texture', 'concave points error', 'mean symmetry', 'm

## 4. Cross-Validation and Feature-Selection Stability

The single 80/20 train-test experiment above is retained as the fixed-split analysis.  
The following sections add cross-validation and stability analyses to strengthen the breast-cancer experiment.

The purpose is to assess whether the observed ANOVA/DODA feature-selection and predictive results persist across different data partitions rather than relying only on `random_state=42`.

**Important:** feature selection is refit inside every training fold. Validation/test folds are not used to determine ANOVA scores or DODA-selected features.


In [30]:
# =============================================================================
# STEP 21: CROSS-VALIDATION CONFIGURATION
# =============================================================================

import numpy as np
import pandas as pd

from itertools import combinations

from sklearn.model_selection import StratifiedKFold, RepeatedStratifiedKFold
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score
)

from doda import DODASelector
from doda.adapters.sklearn_adapter import SklearnAdapter
from doda.knowledge.providers import JSONProvider
from doda.fusion import HadamardFusion


# -------------------------------------------------------------------------
# Research design
# -------------------------------------------------------------------------

TOP_K_VALUES = [5, 10, 15, 20]

N_SPLITS = 5
N_REPEATS = 5
RANDOM_STATE = 42

WEIGHTS_FILE = "../../../config/clinical_weights/breast_cancer.json"

print("=" * 80)
print("BREAST CANCER CROSS-VALIDATION CONFIGURATION")
print("=" * 80)

print("Top-K values:", TOP_K_VALUES)
print("5-fold CV:", N_SPLITS)
print("Repeated CV:", f"{N_SPLITS} folds × {N_REPEATS} repeats")
print("Total repeated runs:", N_SPLITS * N_REPEATS)
print("Random state:", RANDOM_STATE)
print("Clinical weights:", WEIGHTS_FILE)


BREAST CANCER CROSS-VALIDATION CONFIGURATION
Top-K values: [5, 10, 15, 20]
5-fold CV: 5
Repeated CV: 5 folds × 5 repeats
Total repeated runs: 25
Random state: 42
Clinical weights: ../../../config/clinical_weights/breast_cancer.json


In [31]:
# =============================================================================
# STEP 22: 5-FOLD CV PREDICTIVE PERFORMANCE
# ANOVA vs ANOVA + DODA
# =============================================================================
# This section extends the breast-cancer experiment with cross-validation.
# Unlike the original template, it does NOT perform imputation because the
# breast-cancer processed dataset is checked for missing values above.
#
# Feature selection is fitted inside each training fold.

cv_models = {
    "Logistic Regression": LogisticRegression(
        max_iter=1000,
        class_weight="balanced",
        random_state=RANDOM_STATE
    ),

    "Random Forest": RandomForestClassifier(
        n_estimators=100,
        class_weight="balanced",
        random_state=RANDOM_STATE
    ),

    "XGBoost": XGBClassifier(
        random_state=RANDOM_STATE,
        eval_metric="logloss",
        scale_pos_weight=(
            (y == 0).sum() / (y == 1).sum()
        )
    )
}

cv = StratifiedKFold(
    n_splits=N_SPLITS,
    shuffle=True,
    random_state=RANDOM_STATE
)

cv_results = []

for method in ["ANOVA", "DODA"]:

    print("\n" + "=" * 80)
    print(f"METHOD: {method}")
    print("=" * 80)

    for TOP_K in TOP_K_VALUES:

        for model_name, base_model in cv_models.items():

            for fold, (train_idx, val_idx) in enumerate(
                cv.split(X, y),
                start=1
            ):

                X_train_cv = X.iloc[train_idx].copy()
                X_val_cv = X.iloc[val_idx].copy()

                y_train_cv = y.iloc[train_idx].copy()
                y_val_cv = y.iloc[val_idx].copy()

                # ---------------------------------------------------------
                # FEATURE SELECTION
                # ---------------------------------------------------------

                if method == "ANOVA":

                    selector_cv = SelectKBest(
                        score_func=f_classif,
                        k=TOP_K
                    )

                    selector_cv.fit(
                        X_train_cv,
                        y_train_cv
                    )

                    selected_features = (
                        X_train_cv.columns[
                            selector_cv.get_support()
                        ].tolist()
                    )

                else:

                    # ANOVA scores ALL candidate features.
                    # DODA performs final Top-K selection.
                    anova_operator = SklearnAdapter(
                        SelectKBest(
                            score_func=f_classif,
                            k="all"
                        )
                    )

                    provider_cv = JSONProvider(
                        WEIGHTS_FILE
                    )

                    doda_selector_cv = DODASelector(
                        operators=[anova_operator],
                        provider=provider_cv,
                        fusion=HadamardFusion(),
                        top_k=TOP_K
                    )

                    doda_selector_cv.fit(
                        X_train_cv,
                        y_train_cv
                    )

                    selected_features = list(
                        doda_selector_cv.get_selected_features()
                    )

                X_train_selected = X_train_cv[
                    selected_features
                ]

                X_val_selected = X_val_cv[
                    selected_features
                ]

                # ---------------------------------------------------------
                # FRESH MODEL FOR THIS FOLD
                # ---------------------------------------------------------

                model = type(base_model)(
                    **base_model.get_params()
                )

                Xtr = X_train_selected
                Xva = X_val_selected

                # Only Logistic Regression requires scaling.
                if model_name == "Logistic Regression":

                    fold_scaler = StandardScaler()

                    Xtr = fold_scaler.fit_transform(Xtr)
                    Xva = fold_scaler.transform(Xva)

                model.fit(
                    Xtr,
                    y_train_cv
                )

                y_pred = model.predict(Xva)
                y_prob = model.predict_proba(Xva)[:, 1]

                cv_results.append({

                    "Fold": fold,
                    "Method": method,
                    "Top_K": TOP_K,
                    "Model": model_name,

                    "Accuracy": accuracy_score(
                        y_val_cv,
                        y_pred
                    ),

                    "Precision": precision_score(
                        y_val_cv,
                        y_pred,
                        zero_division=0
                    ),

                    "Recall": recall_score(
                        y_val_cv,
                        y_pred,
                        zero_division=0
                    ),

                    "F1": f1_score(
                        y_val_cv,
                        y_pred,
                        zero_division=0
                    ),

                    "ROC_AUC": roc_auc_score(
                        y_val_cv,
                        y_prob
                    ),

                    "Selected_Features": ", ".join(
                        selected_features
                    )
                })

cv_results_df = pd.DataFrame(cv_results)

print("\nCompleted 5-fold CV.")
print("Rows:", len(cv_results_df))

display(cv_results_df.head())



METHOD: ANOVA

METHOD: DODA
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x000001A142B725D0>
DODA Pipeline Started
Running: SelectKBest
DODA feature order:
['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension']
Number of features: 30
Number of scores: 30
Operator Scores:
{'SelectKBest_1': {'mean radius': np.float64(511.49698182786676), 'mean texture': np.float64(110.04158023533923), 'mean perimeter': np.float64(546.5956230324064), 'mean area': np.float64(448.7907742833436), 'mean smoothness': np.float64(64.24216496706457), 'mean compactness': np.float64(240.39624526898402), 'mean concavity': np.float64(409.96209156982934), 'mean concave points': np.float64(655.9966827826175), 'mean symmetry': np.float64(53.97686576503337), 'mean fractal dimension': np.float64(0.6065521620023456), 'radius error': np.float6

,Fold,Method,Top_K,Model,Accuracy,Precision,Recall,F1,ROC_AUC,Selected_Features
0,1,ANOVA,5,Logistic Regression,0.938596,1.000000,0.901408,0.948148,0.997707,"mean perimeter, mean concave points, worst rad..."
1,2,ANOVA,5,Logistic Regression,0.921053,0.918919,0.957746,0.937931,0.978709,"mean perimeter, mean concave points, worst rad..."
2,3,ANOVA,5,Logistic Regression,0.947368,0.958333,0.958333,0.958333,0.977844,"mean perimeter, mean concave points, worst rad..."
3,4,ANOVA,5,Logistic Regression,0.956140,1.000000,0.930556,0.964029,0.993717,"mean perimeter, mean concave points, worst rad..."
4,5,ANOVA,5,Logistic Regression,0.964602,0.958904,0.985915,0.972222,0.996982,"mean concave points, worst radius, worst perim..."


In [32]:
# =============================================================================
# STEP 23: 5-FOLD CV SUMMARY — MEAN ± STD
# =============================================================================

cv_summary_df = (
    cv_results_df
    .groupby(
        ["Method", "Top_K", "Model"]
    )
    .agg(
        Accuracy_Mean=("Accuracy", "mean"),
        Accuracy_STD=("Accuracy", "std"),

        Precision_Mean=("Precision", "mean"),
        Precision_STD=("Precision", "std"),

        Recall_Mean=("Recall", "mean"),
        Recall_STD=("Recall", "std"),

        F1_Mean=("F1", "mean"),
        F1_STD=("F1", "std"),

        ROC_AUC_Mean=("ROC_AUC", "mean"),
        ROC_AUC_STD=("ROC_AUC", "std")
    )
    .reset_index()
)

print("=" * 100)
print("5-FOLD CV PERFORMANCE SUMMARY")
print("=" * 100)

display(cv_summary_df.round(4))


# -------------------------------------------------------------------------
# Save
# -------------------------------------------------------------------------

os.makedirs(
    "../../../results/breast_cancer/cv",
    exist_ok=True
)

cv_results_df.to_csv(
    "../../../results/breast_cancer/cv/anova_doda_5fold_results.csv",
    index=False
)

cv_summary_df.to_csv(
    "../../../results/breast_cancer/cv/anova_doda_5fold_summary.csv",
    index=False
)


5-FOLD CV PERFORMANCE SUMMARY


,Method,Top_K,Model,Accuracy_Mean,Accuracy_STD,Precision_Mean,Precision_STD,Recall_Mean,Recall_STD,F1_Mean,F1_STD,ROC_AUC_Mean,ROC_AUC_STD
0,ANOVA,5,Logistic Regression,0.9456,0.0168,0.9672,0.0340,0.9468,0.0320,0.9561,0.0134,0.9890,0.0099
1,ANOVA,5,Random Forest,0.9385,0.0137,0.9586,0.0279,0.9440,0.0327,0.9506,0.0113,0.9829,0.0119
2,ANOVA,5,XGBoost,0.9350,0.0158,0.9559,0.0326,0.9412,0.0249,0.9479,0.0120,0.9748,0.0192
3,ANOVA,10,Logistic Regression,0.9438,0.0132,0.9723,0.0259,0.9384,0.0416,0.9542,0.0120,0.9895,0.0097
4,ANOVA,10,Random Forest,0.9438,0.0170,0.9613,0.0271,0.9496,0.0287,0.9549,0.0137,0.9845,0.0121
5,ANOVA,10,XGBoost,0.9403,0.0209,0.9606,0.0175,0.9441,0.0354,0.9518,0.0175,0.9783,0.0140
6,ANOVA,15,Logistic Regression,0.9473,0.0223,0.9721,0.0280,0.9441,0.0355,0.9573,0.0180,0.9918,0.0068
7,ANOVA,15,Random Forest,0.9490,0.0169,0.9643,0.0277,0.9552,0.0316,0.9592,0.0139,0.9878,0.0092
8,ANOVA,15,XGBoost,0.9491,0.0227,0.9582,0.0172,0.9608,0.0229,0.9594,0.0182,0.9856,0.0066
9,ANOVA,20,Logistic Regression,0.9772,0.0100,0.9837,0.0176,0.9804,0.0273,0.9817,0.0083,0.9963,0.0061


In [33]:
# =============================================================================
# STEP 24: 5×5 REPEATED STRATIFIED CV
# ANOVA vs ANOVA + DODA
# =============================================================================
# 5 folds × 5 repeats = 25 independent train/validation partitions.
# Feature selection is refit within every run.
#
# No imputation is performed because the dataset was verified to contain
# no missing values.
#
# The repeated-CV results are kept in a separate dataframe so that they
# do not overwrite the fixed-split baseline results_df.

repeated_cv = RepeatedStratifiedKFold(
    n_splits=N_SPLITS,
    n_repeats=N_REPEATS,
    random_state=RANDOM_STATE
)

repeated_cv_results = []

for run_id, (train_idx, val_idx) in enumerate(
    repeated_cv.split(X, y),
    start=1
):

    print("=" * 70)
    print(f"RUN {run_id}/{N_SPLITS * N_REPEATS}")
    print("=" * 70)

    X_train_rep = X.iloc[train_idx].copy()
    X_val_rep = X.iloc[val_idx].copy()

    y_train_rep = y.iloc[train_idx].copy()
    y_val_rep = y.iloc[val_idx].copy()

    for TOP_K in TOP_K_VALUES:

        # ---------------------------------------------------------------
        # ANOVA baseline
        # ---------------------------------------------------------------

        anova_selector_rep = SelectKBest(
            score_func=f_classif,
            k=TOP_K
        )

        anova_selector_rep.fit(
            X_train_rep,
            y_train_rep
        )

        anova_features_rep = (
            X_train_rep.columns[
                anova_selector_rep.get_support()
            ].tolist()
        )

        # ---------------------------------------------------------------
        # DODA
        # ---------------------------------------------------------------

        anova_operator_rep = SklearnAdapter(
            SelectKBest(
                score_func=f_classif,
                k="all"
            )
        )

        provider_rep = JSONProvider(
            WEIGHTS_FILE
        )

        doda_selector_rep = DODASelector(
            operators=[anova_operator_rep],
            provider=provider_rep,
            fusion=HadamardFusion(),
            top_k=TOP_K
        )

        doda_selector_rep.fit(
            X_train_rep,
            y_train_rep
        )

        doda_features_rep = list(
            doda_selector_rep.get_selected_features()
        )

        # ---------------------------------------------------------------
        # Evaluate both feature-selection conditions with each model
        # ---------------------------------------------------------------

        for model_name, model_template in cv_models.items():

            # =========================
            # ANOVA
            # =========================

            model = type(model_template)(
                **model_template.get_params()
            )

            Xtr = X_train_rep[
                anova_features_rep
            ]

            Xva = X_val_rep[
                anova_features_rep
            ]

            if model_name == "Logistic Regression":

                rep_scaler = StandardScaler()

                Xtr = rep_scaler.fit_transform(Xtr)
                Xva = rep_scaler.transform(Xva)

            model.fit(
                Xtr,
                y_train_rep
            )

            y_pred = model.predict(Xva)
            y_prob = model.predict_proba(Xva)[:, 1]

            repeated_cv_results.append({

                "Run": run_id,
                "Method": "ANOVA",
                "Top_K": TOP_K,
                "Model": model_name,

                "Accuracy": accuracy_score(
                    y_val_rep,
                    y_pred
                ),

                "Precision": precision_score(
                    y_val_rep,
                    y_pred,
                    zero_division=0
                ),

                "Recall": recall_score(
                    y_val_rep,
                    y_pred,
                    zero_division=0
                ),

                "F1": f1_score(
                    y_val_rep,
                    y_pred,
                    zero_division=0
                ),

                "ROC_AUC": roc_auc_score(
                    y_val_rep,
                    y_prob
                ),

                "Selected_Features": ", ".join(
                    anova_features_rep
                )
            })

            # =========================
            # DODA
            # =========================

            model = type(model_template)(
                **model_template.get_params()
            )

            Xtr = X_train_rep[
                doda_features_rep
            ]

            Xva = X_val_rep[
                doda_features_rep
            ]

            if model_name == "Logistic Regression":

                rep_scaler = StandardScaler()

                Xtr = rep_scaler.fit_transform(Xtr)
                Xva = rep_scaler.transform(Xva)

            model.fit(
                Xtr,
                y_train_rep
            )

            y_pred = model.predict(Xva)
            y_prob = model.predict_proba(Xva)[:, 1]

            repeated_cv_results.append({

                "Run": run_id,
                "Method": "DODA",
                "Top_K": TOP_K,
                "Model": model_name,

                "Accuracy": accuracy_score(
                    y_val_rep,
                    y_pred
                ),

                "Precision": precision_score(
                    y_val_rep,
                    y_pred,
                    zero_division=0
                ),

                "Recall": recall_score(
                    y_val_rep,
                    y_pred,
                    zero_division=0
                ),

                "F1": f1_score(
                    y_val_rep,
                    y_pred,
                    zero_division=0
                ),

                "ROC_AUC": roc_auc_score(
                    y_val_rep,
                    y_prob
                ),

                "Selected_Features": ", ".join(
                    doda_features_rep
                )
            })

repeated_cv_results_df = pd.DataFrame(
    repeated_cv_results
)

print("\nCompleted repeated CV.")
print("Rows:", len(repeated_cv_results_df))

display(
    repeated_cv_results_df.head()
)


RUN 1/25
Provider inside knowledge engine: <doda.knowledge.providers.json_provider.JSONProvider object at 0x000001A142B7ADB0>
DODA Pipeline Started
Running: SelectKBest
DODA feature order:
['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension']
Number of features: 30
Number of scores: 30
Operator Scores:
{'SelectKBest_1': {'mean radius': np.float64(511.49698182786676), 'mean texture': np.float64(110.04158023533923), 'mean perimeter': np.float64(546.5956230324064), 'mean area': np.float64(448.7907742833436), 'mean smoothness': np.float64(64.24216496706457), 'mean compactness': np.float64(240.39624526898402), 'mean concavity': np.float64(409.96209156982934), 'mean concave points': np.float64(655.9966827826175), 'mean symmetry': np.float64(53.97686576503337), 'mean fractal dimension': np.float64(0.6065521620023456), 'radius error': np.float64(207.36130359442387

,Run,Method,Top_K,Model,Accuracy,Precision,Recall,F1,ROC_AUC,Selected_Features
0,1,ANOVA,5,Logistic Regression,0.938596,1.000000,0.901408,0.948148,0.997707,"mean perimeter, mean concave points, worst rad..."
1,1,DODA,5,Logistic Regression,0.938596,1.000000,0.901408,0.948148,0.997707,"worst concave points, worst perimeter, worst r..."
2,1,ANOVA,5,Random Forest,0.938596,0.984848,0.915493,0.948905,0.995087,"mean perimeter, mean concave points, worst rad..."
3,1,DODA,5,Random Forest,0.956140,1.000000,0.929577,0.963504,0.996233,"worst concave points, worst perimeter, worst r..."
4,1,ANOVA,5,XGBoost,0.938596,0.984848,0.915493,0.948905,0.996069,"mean perimeter, mean concave points, worst rad..."


In [34]:
# =============================================================================
# STEP 25: REPEATED CV PERFORMANCE SUMMARY — MEAN ± STD
# =============================================================================

repeated_cv_summary_df = (
    repeated_cv_results_df
    .groupby(
        ["Method", "Top_K", "Model"]
    )
    .agg(
        Accuracy_Mean=("Accuracy", "mean"),
        Accuracy_STD=("Accuracy", "std"),

        Precision_Mean=("Precision", "mean"),
        Precision_STD=("Precision", "std"),

        Recall_Mean=("Recall", "mean"),
        Recall_STD=("Recall", "std"),

        F1_Mean=("F1", "mean"),
        F1_STD=("F1", "std"),

        ROC_AUC_Mean=("ROC_AUC", "mean"),
        ROC_AUC_STD=("ROC_AUC", "std")
    )
    .reset_index()
)

print("=" * 100)
print("5×5 REPEATED STRATIFIED CV PERFORMANCE SUMMARY")
print("=" * 100)

display(
    repeated_cv_summary_df.round(4)
)

repeated_cv_results_df.to_csv(
    "../../../results/breast_cancer/cv/anova_doda_5x5_repeated_results.csv",
    index=False
)

repeated_cv_summary_df.to_csv(
    "../../../results/breast_cancer/cv/anova_doda_5x5_repeated_summary.csv",
    index=False
)


5×5 REPEATED STRATIFIED CV PERFORMANCE SUMMARY


,Method,Top_K,Model,Accuracy_Mean,Accuracy_STD,Precision_Mean,Precision_STD,Recall_Mean,Recall_STD,F1_Mean,F1_STD,ROC_AUC_Mean,ROC_AUC_STD
0,ANOVA,5,Logistic Regression,0.9469,0.0189,0.9698,0.0226,0.9456,0.0307,0.9570,0.0159,0.9881,0.0069
1,ANOVA,5,Random Forest,0.9392,0.0179,0.9575,0.0271,0.9462,0.0279,0.9512,0.0144,0.9821,0.0095
2,ANOVA,5,XGBoost,0.9378,0.0198,0.9542,0.0262,0.9473,0.0290,0.9502,0.0161,0.9775,0.0131
3,ANOVA,10,Logistic Regression,0.9427,0.0170,0.9711,0.0204,0.9372,0.0309,0.9534,0.0145,0.9883,0.0073
4,ANOVA,10,Random Forest,0.9469,0.0169,0.9620,0.0254,0.9540,0.0260,0.9575,0.0135,0.9848,0.0089
5,ANOVA,10,XGBoost,0.9445,0.0160,0.9572,0.0243,0.9552,0.0255,0.9557,0.0128,0.9789,0.0110
6,ANOVA,15,Logistic Regression,0.9480,0.0164,0.9723,0.0202,0.9445,0.0230,0.9579,0.0133,0.9916,0.0046
7,ANOVA,15,Random Forest,0.9483,0.0203,0.9633,0.0288,0.9552,0.0271,0.9587,0.0163,0.9872,0.0082
8,ANOVA,15,XGBoost,0.9466,0.0179,0.9566,0.0221,0.9591,0.0263,0.9574,0.0146,0.9883,0.0067
9,ANOVA,20,Logistic Regression,0.9768,0.0134,0.9812,0.0147,0.9821,0.0178,0.9815,0.0107,0.9955,0.0048


## 5. Feature-Selection Stability and ANOVA–DODA Agreement

The next analyses use the **25 repeated-CV runs** to examine how consistently features are selected.

We distinguish:

- **Selection frequency:** how often a feature is selected.
- **Jaccard similarity:** overlap between feature sets.
- **Kuncheva index:** chance-adjusted agreement for fixed-size feature subsets.
- **ANOVA–DODA agreement:** whether DODA changes the selected set relative to the statistical baseline.


In [35]:
# =============================================================================
# STEP 26: PREPARE STABILITY DATAFRAME
# =============================================================================

stability_df = repeated_cv_results_df[
    [
        "Run",
        "Method",
        "Top_K",
        "Selected_Features"
    ]
].copy()

stability_df["Feature_Set"] = (
    stability_df["Selected_Features"]
    .apply(
        lambda x: {
            feature.strip()
            for feature in x.split(",")
            if feature.strip()
        }
    )
)

feature_names = X.columns.tolist()
N_FEATURES = len(feature_names)

print("=" * 80)
print("STABILITY ANALYSIS PREPARATION")
print("=" * 80)

print("Rows:", stability_df.shape[0])
print("Methods:", stability_df["Method"].unique())
print("Top-K:", sorted(stability_df["Top_K"].unique()))
print("Total predictors:", N_FEATURES)

display(
    stability_df[
        [
            "Run",
            "Method",
            "Top_K",
            "Feature_Set"
        ]
    ].head()
)


STABILITY ANALYSIS PREPARATION
Rows: 600
Methods: <StringArray>
['ANOVA', 'DODA']
Length: 2, dtype: str
Top-K: [np.int64(5), np.int64(10), np.int64(15), np.int64(20)]
Total predictors: 30


,Run,Method,Top_K,Feature_Set
0,1,ANOVA,5,"{worst perimeter, mean perimeter, mean concave..."
1,1,DODA,5,"{worst perimeter, mean concave points, worst r..."
2,1,ANOVA,5,"{worst perimeter, mean perimeter, mean concave..."
3,1,DODA,5,"{worst perimeter, mean concave points, worst r..."
4,1,ANOVA,5,"{worst perimeter, mean perimeter, mean concave..."


In [36]:
# =============================================================================
# STEP 27: TEST 1 — FEATURE SELECTION FREQUENCY
# =============================================================================

selection_frequency = []

for method in sorted(
    stability_df["Method"].unique()
):

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ]

        total_runs = len(subset)

        all_features = sorted(
            set().union(
                *subset["Feature_Set"]
            )
        )

        for feature in all_features:

            selected_count = sum(
                feature in feature_set
                for feature_set in subset["Feature_Set"]
            )

            selection_frequency.append({

                "Method": method,

                "Top_K": top_k,

                "Feature": feature,

                "Selected_Count": selected_count,

                "Total_Runs": total_runs,

                "Selection_Frequency_Percentage":
                    (
                        selected_count /
                        total_runs
                    ) * 100
            })

selection_frequency_df = pd.DataFrame(
    selection_frequency
)

selection_frequency_df = (
    selection_frequency_df
    .sort_values(
        by=[
            "Method",
            "Top_K",
            "Selection_Frequency_Percentage"
        ],
        ascending=[
            True,
            True,
            False
        ]
    )
    .reset_index(drop=True)
)

display(selection_frequency_df)


,Method,Top_K,Feature,Selected_Count,Total_Runs,Selection_Frequency_Percentage
0,ANOVA,5,mean concave points,75,75,100.0
1,ANOVA,5,worst concave points,75,75,100.0
2,ANOVA,5,worst perimeter,75,75,100.0
3,ANOVA,5,worst radius,75,75,100.0
4,ANOVA,5,mean perimeter,69,75,92.0
...,...,...,...,...,...,...
100,DODA,20,worst smoothness,75,75,100.0
101,DODA,20,worst symmetry,75,75,100.0
102,DODA,20,worst texture,75,75,100.0
103,DODA,20,concave points error,72,75,96.0


In [37]:
# =============================================================================
# STEP 28: TOP FREQUENTLY SELECTED FEATURES
# =============================================================================

top_frequency_features = (
    selection_frequency_df
    .groupby(
        ["Method", "Top_K"],
        group_keys=False
    )
    .head(10)
    .reset_index(drop=True)
)

print("=" * 80)
print("TOP FREQUENTLY SELECTED FEATURES")
print("=" * 80)

display(top_frequency_features)


TOP FREQUENTLY SELECTED FEATURES


,Method,Top_K,Feature,Selected_Count,Total_Runs,Selection_Frequency_Percentage
0,ANOVA,5,mean concave points,75,75,100.0
1,ANOVA,5,worst concave points,75,75,100.0
2,ANOVA,5,worst perimeter,75,75,100.0
3,ANOVA,5,worst radius,75,75,100.0
4,ANOVA,5,mean perimeter,69,75,92.0
...,...,...,...,...,...,...
68,DODA,20,mean perimeter,75,75,100.0
69,DODA,20,mean radius,75,75,100.0
70,DODA,20,mean texture,75,75,100.0
71,DODA,20,perimeter error,75,75,100.0


In [38]:
# =============================================================================
# STEP 29: TEST 2 — PAIRWISE JACCARD SIMILARITY
# =============================================================================

jaccard_results = []

for method in sorted(
    stability_df["Method"].unique()
):

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ].reset_index(drop=True)

        feature_sets = (
            subset["Feature_Set"]
            .tolist()
        )

        pairwise_scores = []

        for set_a, set_b in combinations(
            feature_sets,
            2
        ):

            intersection_size = len(
                set_a.intersection(set_b)
            )

            union_size = len(
                set_a.union(set_b)
            )

            jaccard_score = (
                intersection_size / union_size
                if union_size
                else 1.0
            )

            pairwise_scores.append(
                jaccard_score
            )

        if pairwise_scores:

            jaccard_results.append({

                "Method": method,

                "Top_K": top_k,

                "Mean_Jaccard": np.mean(
                    pairwise_scores
                ),

                "Std_Jaccard": np.std(
                    pairwise_scores,
                    ddof=1
                ),

                "Minimum_Jaccard": np.min(
                    pairwise_scores
                ),

                "Maximum_Jaccard": np.max(
                    pairwise_scores
                )
            })

jaccard_df = pd.DataFrame(
    jaccard_results
)

print("=" * 80)
print("PAIRWISE JACCARD SIMILARITY")
print("=" * 80)

display(jaccard_df.round(4))


PAIRWISE JACCARD SIMILARITY


,Method,Top_K,Mean_Jaccard,Std_Jaccard,Minimum_Jaccard,Maximum_Jaccard
0,ANOVA,5,0.9492,0.1198,0.6667,1.0
1,ANOVA,10,1.0000,0.0000,1.0000,1.0
2,ANOVA,15,1.0000,0.0000,1.0000,1.0
3,ANOVA,20,0.9855,0.0342,0.9048,1.0
4,DODA,5,0.9503,0.1188,0.6667,1.0
5,DODA,10,1.0000,0.0000,1.0000,1.0
6,DODA,15,1.0000,0.0000,1.0000,1.0
7,DODA,20,0.9926,0.0255,0.9048,1.0


In [39]:
# =============================================================================
# STEP 30: TEST 3 — KUNCHEVA INDEX
# =============================================================================

kuncheva_results = []

for method in sorted(
    stability_df["Method"].unique()
):

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        subset = stability_df[
            (stability_df["Method"] == method) &
            (stability_df["Top_K"] == top_k)
        ].reset_index(drop=True)

        feature_sets = (
            subset["Feature_Set"]
            .tolist()
        )

        for i in range(
            len(feature_sets)
        ):

            for j in range(
                i + 1,
                len(feature_sets)
            ):

                set_a = feature_sets[i]
                set_b = feature_sets[j]

                intersection = len(
                    set_a.intersection(set_b)
                )

                denominator = (
                    top_k *
                    (
                        N_FEATURES -
                        top_k
                    )
                )

                if denominator == 0:

                    kuncheva = np.nan

                else:

                    kuncheva = (
                        (
                            intersection *
                            N_FEATURES
                        )
                        -
                        (
                            top_k ** 2
                        )
                    ) / denominator

                kuncheva_results.append({

                    "Method": method,

                    "Top_K": top_k,

                    "Run_A": subset.loc[
                        i,
                        "Run"
                    ],

                    "Run_B": subset.loc[
                        j,
                        "Run"
                    ],

                    "Intersection": intersection,

                    "Kuncheva_Index": kuncheva
                })

kuncheva_df = pd.DataFrame(
    kuncheva_results
)

kuncheva_summary = (
    kuncheva_df
    .groupby(
        ["Method", "Top_K"]
    )["Kuncheva_Index"]
    .agg(
        Mean_Kuncheva="mean",
        Std_Kuncheva="std",
        Min_Kuncheva="min",
        Max_Kuncheva="max"
    )
    .reset_index()
)

print("=" * 80)
print("KUNCHEVA INDEX SUMMARY")
print("=" * 80)

display(
    kuncheva_summary.round(4)
)


KUNCHEVA INDEX SUMMARY


,Method,Top_K,Mean_Kuncheva,Std_Kuncheva,Min_Kuncheva,Max_Kuncheva
0,ANOVA,5,0.9634,0.0863,0.76,1.0
1,ANOVA,10,1.0000,0.0000,1.00,1.0
2,ANOVA,15,1.0000,0.0000,1.00,1.0
3,ANOVA,20,0.9771,0.0539,0.85,1.0
4,DODA,5,0.9642,0.0855,0.76,1.0
5,DODA,10,1.0000,0.0000,1.00,1.0
6,DODA,15,1.0000,0.0000,1.00,1.0
7,DODA,20,0.9883,0.0402,0.85,1.0


In [40]:
# =============================================================================
# STEP 31: TEST 4 — ANOVA vs DODA FEATURE-SET AGREEMENT
# =============================================================================
# This compares feature-set membership within the SAME repeated-CV run
# and Top-K value.

agreement_results = []

for run in sorted(
    stability_df["Run"].unique()
):

    for top_k in sorted(
        stability_df["Top_K"].unique()
    ):

        anova_rows = stability_df[
            (stability_df["Run"] == run) &
            (stability_df["Method"] == "ANOVA") &
            (stability_df["Top_K"] == top_k)
        ]

        doda_rows = stability_df[
            (stability_df["Run"] == run) &
            (stability_df["Method"] == "DODA") &
            (stability_df["Top_K"] == top_k)
        ]

        if (
            anova_rows.empty or
            doda_rows.empty
        ):
            continue

        anova_set = (
            anova_rows.iloc[0]["Feature_Set"]
        )

        doda_set = (
            doda_rows.iloc[0]["Feature_Set"]
        )

        intersection_size = len(
            anova_set.intersection(
                doda_set
            )
        )

        union_size = len(
            anova_set.union(
                doda_set
            )
        )

        jaccard_agreement = (
            intersection_size / union_size
            if union_size
            else 1.0
        )

        agreement_results.append({

            "Run": run,

            "Top_K": top_k,

            "ANOVA_Features":
                ", ".join(
                    sorted(anova_set)
                ),

            "DODA_Features":
                ", ".join(
                    sorted(doda_set)
                ),

            "Common_Feature_Count":
                intersection_size,

            "ANOVA_Feature_Count":
                len(anova_set),

            "DODA_Feature_Count":
                len(doda_set),

            "Same_Feature_Set":
                anova_set == doda_set,

            "Jaccard_Agreement":
                jaccard_agreement,

            "Features_Added_by_DODA":
                ", ".join(
                    sorted(
                        doda_set -
                        anova_set
                    )
                ),

            "Features_Removed_by_DODA":
                ", ".join(
                    sorted(
                        anova_set -
                        doda_set
                    )
                )
        })

agreement_df = pd.DataFrame(
    agreement_results
)

print("=" * 100)
print("ANOVA vs DODA FEATURE-SET AGREEMENT")
print("=" * 100)

display(agreement_df)


ANOVA vs DODA FEATURE-SET AGREEMENT


,Run,Top_K,ANOVA_Features,DODA_Features,Common_Feature_Count,ANOVA_Feature_Count,DODA_Feature_Count,Same_Feature_Set,Jaccard_Agreement,Features_Added_by_DODA,Features_Removed_by_DODA
0,1,5,"mean concave points, mean perimeter, worst con...","mean concave points, worst area, worst concave...",4,5,5,False,0.666667,worst area,mean perimeter
1,1,10,"mean area, mean concave points, mean concavity...","mean area, mean concave points, mean concavity...",10,10,10,True,1.000000,,
2,1,15,"area error, mean area, mean compactness, mean ...","area error, mean area, mean compactness, mean ...",15,15,15,True,1.000000,,
3,1,20,"area error, concave points error, mean area, m...","area error, concave points error, mean area, m...",20,20,20,True,1.000000,,
4,2,5,"mean concave points, mean perimeter, worst con...","mean concave points, worst area, worst concave...",4,5,5,False,0.666667,worst area,mean perimeter
...,...,...,...,...,...,...,...,...,...,...,...
95,24,20,"area error, concave points error, mean area, m...","area error, concave points error, mean area, m...",20,20,20,True,1.000000,,
96,25,5,"mean concave points, mean perimeter, worst con...","mean concave points, worst area, worst concave...",4,5,5,False,0.666667,worst area,mean perimeter
97,25,10,"mean area, mean concave points, mean concavity...","mean area, mean concave points, mean concavity...",10,10,10,True,1.000000,,
98,25,15,"area error, mean area, mean compactness, mean ...","area error, mean area, mean compactness, mean ...",15,15,15,True,1.000000,,


In [41]:
# =============================================================================
# STEP 32: SUMMARY OF ANOVA vs DODA AGREEMENT
# =============================================================================

agreement_summary_df = (
    agreement_df
    .groupby("Top_K")
    .agg(
        Same_Feature_Set_Percentage=(
            "Same_Feature_Set",
            lambda x: x.mean() * 100
        ),

        Mean_Jaccard_Agreement=(
            "Jaccard_Agreement",
            "mean"
        ),

        Std_Jaccard_Agreement=(
            "Jaccard_Agreement",
            "std"
        ),

        Mean_Common_Features=(
            "Common_Feature_Count",
            "mean"
        )
    )
    .reset_index()
)

print("=" * 100)
print("FINAL ANOVA vs DODA FEATURE-SET AGREEMENT SUMMARY")
print("=" * 100)

display(
    agreement_summary_df.round(4)
)


# -------------------------------------------------------------------------
# Save stability outputs
# -------------------------------------------------------------------------

os.makedirs(
    "../../../results/breast_cancer/stability",
    exist_ok=True
)

selection_frequency_df.to_csv(
    "../../../results/breast_cancer/stability/feature_selection_frequency.csv",
    index=False
)

jaccard_df.to_csv(
    "../../../results/breast_cancer/stability/jaccard_similarity.csv",
    index=False
)

kuncheva_df.to_csv(
    "../../../results/breast_cancer/stability/kuncheva_pairwise.csv",
    index=False
)

kuncheva_summary.to_csv(
    "../../../results/breast_cancer/stability/kuncheva_summary.csv",
    index=False
)

agreement_df.to_csv(
    "../../../results/breast_cancer/stability/anova_doda_agreement.csv",
    index=False
)

agreement_summary_df.to_csv(
    "../../../results/breast_cancer/stability/anova_doda_agreement_summary.csv",
    index=False
)


FINAL ANOVA vs DODA FEATURE-SET AGREEMENT SUMMARY


,Top_K,Same_Feature_Set_Percentage,Mean_Jaccard_Agreement,Std_Jaccard_Agreement,Mean_Common_Features
0,5,12.0,0.7067,0.1106,4.12
1,10,100.0,1.0000,0.0000,10.00
2,15,100.0,1.0000,0.0000,15.00
3,20,96.0,0.9962,0.0190,19.96


In [42]:
# =============================================================================
# STEP 33: FINAL NOTEBOOK CONSISTENCY CHECK
# =============================================================================

print("=" * 80)
print("BREAST CANCER ANOVA + DODA NOTEBOOK SANITY CHECK")
print("=" * 80)

print("Dataset shape             :", df.shape)
print("Predictor count           :", X.shape[1])
print("Target column             :", target_column)
print("Missing values in X       :", int(X.isnull().sum().sum()))
print("Fixed-split Top-K values  :", k_values)
print("CV Top-K values           :", TOP_K_VALUES)
print("Repeated-CV runs          :", N_SPLITS * N_REPEATS)
print("Baseline result rows      :", len(results_df))
print("DODA result rows          :", len(doda_results_df))
print("5-fold CV result rows     :", len(cv_results_df))
print("5×5 CV result rows        :", len(repeated_cv_results_df))
print("Stability methods         :", stability_df["Method"].unique())

assert target_column == "target"
assert X.shape[1] == len(X.columns)
assert X.isnull().sum().sum() == 0
assert TOP_K_VALUES == [5, 10, 15, 20]
assert all(
    k <= X.shape[1]
    for k in TOP_K_VALUES
)

print("\nCore notebook dependencies are defined.")
print("Run this notebook from the top using Run All for a clean state.")


BREAST CANCER ANOVA + DODA NOTEBOOK SANITY CHECK
Dataset shape             : (569, 31)
Predictor count           : 30
Target column             : target
Missing values in X       : 0
Fixed-split Top-K values  : [5, 10, 15, 20]
CV Top-K values           : [5, 10, 15, 20]
Repeated-CV runs          : 25
Baseline result rows      : 12
DODA result rows          : 12
5-fold CV result rows     : 120
5×5 CV result rows        : 600
Stability methods         : <StringArray>
['ANOVA', 'DODA']
Length: 2, dtype: str

Core notebook dependencies are defined.
Run this notebook from the top using Run All for a clean state.


# 6. ANOVA + DODA: Model Evaluation & Interpretation

---

## 3.1 Overall Predictive Performance

### 📊 Best Overall Results

| Method | Top-K | Model | Accuracy | Precision | Recall | F1-score | ROC-AUC |
|:------|------:|:------|---------:|---------:|-------:|---------:|--------:|
| ANOVA | 10 | **XGBoost** | **0.9737** | 0.9726 | **0.9861** | **0.9793** | 0.9924 |
| ANOVA | 20 | **XGBoost** | 0.9386 | 0.9577 | 0.9444 | 0.9510 | **0.9954** |
| DODA | 20 | **XGBoost** | 0.9474 | 0.9577 | 0.9583 | 0.9583 | **0.9960** |

> **Key observation:**  
> ANOVA + XGBoost at Top-K = 10 achieved the highest ANOVA accuracy and F1-score, while DODA + XGBoost at Top-K = 20 achieved the highest overall ROC-AUC (**0.9960**).

---

### 📈 ANOVA vs DODA: Performance Comparison

#### Top-K = 5

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Overall Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:---------------|
| Logistic Regression | 0.9298 → **0.9386** | 0.9848 → 0.9863 | 0.9028 → **0.9167** | 0.9420 → **0.9496** | 0.9911 → 0.9907 | 🟢 **Improved** |
| Random Forest | 0.9298 → **0.9474** | 0.9571 → **0.9714** | 0.9306 → 0.9306 | 0.9437 → **0.9507** | 0.9883 → 0.9828 | 🟢 **Mostly improved** |
| XGBoost | 0.9386 → 0.9298 | 0.9577 → 0.9452 | 0.9444 → 0.9306 | 0.9510 → 0.9378 | 0.9845 → 0.9815 | 🔴 **Reduced** |

---

#### Top-K = 10

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Overall Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:---------------|
| Logistic Regression | 0.9123 → 0.9123 | 0.9697 → 0.9697 | 0.8889 → 0.8889 | 0.9275 → 0.9275 | 0.9914 → 0.9914 | 🟡 **Maintained** |
| Random Forest | 0.9561 → 0.9474 | 0.9589 → 0.9589 | 0.9722 → 0.9722 | 0.9655 → 0.9583 | 0.9901 → 0.9891 | 🔴 **Reduced** |
| XGBoost | 0.9737 → 0.9649 | 0.9726 → 0.9726 | 0.9861 → 0.9861 | 0.9793 → 0.9726 | 0.9924 → 0.9919 | 🔴 **Reduced** |

---

#### Top-K = 15

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Overall Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:---------------|
| Logistic Regression | 0.9561 → 0.9561 | 0.9855 → 0.9855 | 0.9444 → 0.9444 | 0.9645 → 0.9645 | 0.9914 → 0.9914 | 🟡 **Maintained** |
| Random Forest | 0.9561 → **0.9649** | 0.9718 → 0.9577 | 0.9583 → **0.9722** | 0.9650 → **0.9722** | 0.9917 → 0.9891 | 🟢 **Improved overall** |
| XGBoost | 0.9649 → 0.9649 | 0.9595 → 0.9595 | 0.9861 → 0.9861 | 0.9726 → 0.9726 | 0.9924 → **0.9927** | 🟡 **Maintained** |

---

#### Top-K = 20

| Model | Accuracy | Precision | Recall | F1-score | ROC-AUC | Overall Effect |
|:------|---------:|----------:|-------:|---------:|--------:|:---------------|
| Logistic Regression | 0.9649 → 0.9649 | 0.9857 → 0.9857 | 0.9583 → 0.9583 | 0.9718 → 0.9718 | 0.9950 → 0.9950 | 🟡 **Maintained** |
| Random Forest | 0.9474 → 0.9474 | 0.9714 → 0.9577 | 0.9444 → **0.9583** | 0.9577 → 0.9577 | 0.9937 → **0.9950** | 🟡 **Comparable** |
| XGBoost | 0.9386 → **0.9474** | 0.9577 → 0.9577 | 0.9444 → **0.9583** | 0.9510 → **0.9583** | 0.9954 → **0.9960** | 🟢 **Improved** |

> **Notation:** `ANOVA → DODA`  
> 🟢 Improved &nbsp;&nbsp; 🟡 Maintained / comparable &nbsp;&nbsp; 🔴 Reduced

---

## 3.2 DODA Performance Summary

| Top-K | Logistic Regression | Random Forest | XGBoost |
|:-----:|:-------------------:|:-------------:|:-------:|
| **5** | 🟢 Improved | 🟢 Mostly improved | 🔴 Reduced |
| **10** | 🟡 Maintained | 🔴 Reduced | 🔴 Reduced |
| **15** | 🟡 Maintained | 🟢 Improved overall | 🟡 Maintained |
| **20** | 🟡 Maintained | 🟡 Comparable | 🟢 Improved |

### 🔎 Main Performance Findings

- **Logistic Regression:** DODA maintained performance at Top-K = 10, 15, and 20 and improved performance at Top-K = 5.
- **Random Forest:** DODA improved performance most clearly at Top-K = 5 and 15, while performance was reduced at Top-K = 10.
- **XGBoost:** DODA reduced performance at Top-K = 5 and 10, maintained performance at Top-K = 15, and improved performance at Top-K = 20.
- **Best DODA result:** XGBoost at **Top-K = 20**, achieving **0.9960 ROC-AUC** and **0.9474 accuracy**.

### 🧠 Interpretation

DODA does **not** produce a uniform improvement across all models and feature sizes. Instead, its effect depends on both the classifier and the number of selected features.

This suggests that clinical weighting can modify the feature space without necessarily improving every predictive model. Importantly, DODA generally preserved competitive performance rather than causing a major deterioration in predictive ability.

---

# 3.3 Effect of DODA on Feature Selection

## 📊 ANOVA vs DODA: Top-10 Feature Ranking

| Rank | ANOVA | DODA | Change |
|-----:|:------|:-----|:------|
| 1 | `worst concave points` | `worst concave points` | — |
| 2 | `worst perimeter` | `worst perimeter` | — |
| 3 | `mean concave points` | `worst radius` | ⬆️ |
| 4 | `worst radius` | `mean concave points` | ⬇️ |
| 5 | `mean perimeter` | `worst area` | ⬆️ |
| 6 | `worst area` | `mean perimeter` | ⬇️ |
| 7 | `mean radius` | `mean radius` | — |
| 8 | `mean area` | `mean area` | — |
| 9 | `mean concavity` | `mean concavity` | — |
| 10 | `worst concavity` | `worst concavity` | — |

### 🎯 What Changed?

The **Top-10 feature set itself remained unchanged**.

DODA mainly changed the **internal ranking**:

- `worst radius`: Rank **4 → 3** ⬆️
- `worst area`: Rank **6 → 5** ⬆️
- `mean concave points`: Rank **3 → 4** ⬇️
- `mean perimeter`: Rank **5 → 6** ⬇️

This demonstrates that DODA acts primarily as a **refinement mechanism**, rather than replacing the statistically selected feature space.

---

## 🩺 Clinical Weight Alignment

| Feature | Clinical Weight | ANOVA Rank | DODA Rank | DODA Change |
|:--------|----------------:|-----------:|----------:|:-----------:|
| `worst concave points` | **1.0** | 1 | 1 | — |
| `worst perimeter` | **1.0** | 2 | 2 | — |
| `worst radius` | **1.0** | 4 | 3 | ⬆️ |
| `worst area` | **1.0** | 6 | 5 | ⬆️ |
| `worst concavity` | **1.0** | 10 | 10 | — |
| `mean concave points` | 0.9 | 3 | 4 | ⬇️ |
| `mean perimeter` | 0.9 | 5 | 6 | ⬇️ |
| `mean radius` | 0.9 | 7 | 7 | — |
| `mean area` | 0.9 | 8 | 8 | — |
| `mean concavity` | 0.9 | 9 | 9 | — |

> **Key observation:**  
> Several features with high clinical weights moved upward after DODA weighting. In particular, `worst radius` and `worst area` moved into higher-ranked positions.

---

## 🔄 Additional Ranking Changes

| Feature | ANOVA Rank | DODA Rank | Clinical Weight | Interpretation |
|:--------|-----------:|----------:|----------------:|:---------------|
| `worst compactness` | 12 | 11 | 0.8 | ⬆️ Promoted |
| `mean compactness` | 11 | 12 | 0.7 | ⬇️ Demoted |
| `mean texture` | 20 | 19 | 0.5 | ⬆️ Promoted |
| `concave points error` | 19 | 20 | 0.4 | ⬇️ Demoted |

These changes show that DODA can shift ranking toward features with stronger configured clinical support when their statistical evidence is sufficiently comparable.

---

# 3.4 Overall Interpretation

### 🧬 What ANOVA Contributed

ANOVA identified a compact group of statistically informative features dominated by:

- **Concave-point measurements**
- **Perimeter measurements**
- **Radius measurements**
- **Area measurements**
- **Concavity measurements**

These features consistently appeared near the top of the statistical ranking.

### 🩺 What DODA Added

DODA incorporated clinical knowledge into this statistical ranking.

Rather than replacing the ANOVA-selected feature space, DODA primarily:

1. **Retained the same major feature family**
2. **Reordered clinically weighted features**
3. **Promoted features such as `worst radius` and `worst area`**
4. **Adjusted lower-ranked features according to their clinical weights**
5. **Generally maintained competitive predictive performance**

---

# 3.5 Current Conclusion

> ### 🎯 Main Finding
>
> **DODA provides a clinically informed refinement of the ANOVA feature ranking rather than a complete replacement of the statistically selected feature space.**

| Aspect | Finding |
|:-------|:--------|
| **Predictive performance** | Comparable overall, with improvements in selected model/Top-K combinations |
| **Feature selection** | Major feature family remained stable |
| **Feature ranking** | Clinically weighted features were promoted in several cases |
| **Best DODA result** | XGBoost, Top-K = 20, ROC-AUC = **0.9960** |
| **Strongest ANOVA result** | XGBoost, Top-K = 10, Accuracy = **0.9737**, F1 = **0.9793** |
| **Main contribution of DODA** | Incorporation of clinical knowledge into statistical feature ranking |
| **Overall interpretation** | **Refinement rather than wholesale replacement** |

### 📌 Research Interpretation

The results support the idea that **clinical knowledge can refine statistically selected features** by modifying their ranking toward clinically important variables while generally maintaining competitive predictive performance.

However, the findings should be interpreted within the scope of this experiment:

- The analysis uses **one breast cancer dataset**.
- No statistical significance testing was performed.
- The results do not establish that DODA will generalize to other diseases, datasets, or patient cohorts.
- Further validation on independent datasets is required before making broader claims about DODA.

### ✅ Final Takeaway

**ANOVA provides the statistical foundation, while DODA adds clinical knowledge to refine the feature ranking.**

The results suggest that this combination can preserve strong predictive performance while producing a feature-selection process that is more explicitly informed by clinical priorities.